### ISSUE TITLE
The backend does not explicitly consider individual elements of a 3D array that are zero or one to simplify expressions

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#11412]
I am trying to get the [multibody flexible beam model](http://dx.doi.org/10.1007/s11044-006-9012-8) developed by Luca Viganò and Francesco Schiavo 18 years ago to run in OMC efficiently, so it can be included in MSL 4.1.0. It is a very interesting case of application of matrix equations. 

In this case, matrix equations are extremely useful, because they allow to write the code in a very compact form, as they can be read in the paper. However, it is essential that they are handled efficiently, carrying out all the feasible symbolic manipulations, to avoid the generated code to blow up unnecessarily. In the case of multibody systems, this is particularly true for what concerns structural matrices, which are constant and very sparse. Unfortunately, the current backend is still not up to the task.

Consider the following MWE, which is a runnable excerpt of the flexible beam model:
```modelica
model Test3DArray
  final parameter Real Sbar11[6, 6]=
        [1/3, 0, 0, 1/6, 0, 0;
           0, 0, 0,   0, 0, 0;
           0, 0, 0,   0, 0, 0;
         1/6, 0, 0, 1/3, 0, 0;
           0, 0, 0,   0, 0, 0;
           0, 0, 0,   0, 0, 0];

  parameter Integer N = 3;
  Real B[N, 6, 3*N];
  Real s[:,:] = sum(transpose(B[i, :, :])*Sbar11*B[i,:,:] for i in 1:N);

equation
  B[1,:,:] = [zeros(2, 3*N);
              [zeros(1, 3*(N - 1)), [0, 1, 0]];
              identity(3), zeros(3, 3*(N - 1))];
  B[N,:,:] = [zeros(3, 3*(N - 2)), identity(3), zeros(3, 3);
              zeros(3, 3*(N - 2)), zeros(3, 3), [1, 0, 0; 0, 0, 0; 0, 0, 1]];
  for i in 2:N - 1 loop
    B[i,:,:] = [zeros(6, 3*(i - 2)), identity(6), zeros(6, 3*(N - i))];
  end for;
end Test3DArray;
```
The 2D matrix `s` results from several (constant) matrix multiplications, and is itself a sparse matrix. It is essential that its zero and one elements are symbolically recognized as such. Unfortunately, this doesn't happen. If you run the model and inspect it with the declarative debugger, the element of `B[2,:,:]` show up as 
```modelica
B[2,1,1] := /*Real*/(cat(2, {{}, {}, {}, {}, {}, {}}, {{1, 0, 0, 0, 0, 0}, {0, 1, 0, 0, 0, 0}, {0, 0, 1, 0, 0, 0}, {0, 0, 0, 1, 0, 0}, {0, 0, 0, 0, 1, 0}, {0, 0, 0, 0, 0, 1}}, {{0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {0, 0, 0}})[1, 1])
```
which is not a particularly efficient way to say `B[2,1,1] := 1`, particularly as `N` grows 😅.

As a consequence, also the element of `s` are computed in an inefficient way, e.g.
```modelica
(assign) s[1,1] := 0.3333333333333333 + (0.3333333333333333 * B[2,1,1] + 0.1666666666666667 * B[2,4,1]) * B[2,1,1] + (0.1666666666666667 * B[2,1,1] + 0.3333333333333333 * B[2,4,1]) * B[2,4,1]
```
which is also a very inefficient way to say `s[1,1] = 0.66666666666667`, since `B[2,1,1] = 1` and `B[2,4,1] = 0`.

Not only is the resulting code inefficient in terms of simulation performance, but most importantly it takes forever to build it.

As I understand, this problem can be solved by making sure that the elements of `B[:,:,:]`, which are all scalar constants, mostly zeros and ones, are evaluated as such, and not as individual elements picked out of a large constant matrix. I guess this is an easy fix that you could apply to the old backend, making the handling of such flexible models feasible in OMC.

This is also an interesting use case for the new backend. I'm not sure to which extent it is possible to compute the `s` matrix efficiently and in sparse form (i.e. with symbolic zeros in all the places where the value turns out to be zero) without expanding all the products to their scalar expressions. In fact, I'm a bit skeptical about it. 

If you want to handle dense matrices, there is of course advantage at keeping those matrices together as first-class citizens, but in the case of sparse matrices I can't see how you could do that, since the sparsity structure is not declared explicitly by listing the nonzero elements. Maybe you could reconstruct it a posteriori and then exploit it in some smart way. Or, maybe, one could add some annotations to these matrices, stating that they should be expanded to exploit the sparsity structure, even though in general the backend tries not to expand them. 

I hope the MWE shown above can help you reasoning about it.

As we already discussed a few days ago, I am sure that one interesting application of large array equations in Modelica is to handle FEM models natively, so it is essential that both the code generation process and the generated code can rely on the many zero's and one's found in some of those matrices.

Thanks!

**Comment by casella** [comment]
Adding @bernhardbachmann to the loop for his information.

**Comment by casella** [comment]
Adding @perost

**Comment by perost** [comment]
#11507 fixes the issue with the old backend at least, the `cat` expression is now correctly simplified to a simple literal matrix and you get e.g. `B[2, 1, 1] = 1.0` in the flat model. The evaluation of `cat` is implemented as a generic function and one of the few things shared by the frontends, so this also applies to the old expression simplification module that the old backend uses.

With the new backend the `cat` expression is not simplified since we don't expand `fill` calls, so it ends up as `(cat(2, fill(0, 6, 3 * (i - 2)), identity(6), fill(0, 6, 3 * (3 - i))));` in the flat model. The new backend probably simplifies this at some point, but I'm not sure how to check that.

**Comment by phannebohm** [comment]
> Edit: Just to be clear, this only fixes the issue with `B[2, :, :]`. The binding of `s` still ends up as a huge expression, but I'm not sure what can be done about that.

With the old backend, omc generated the correct equation `b = 5.0` from the following model:
```modelica
model ConstRepl
  Real a[2],b;
equation
  a = vector([ones(1), fill(2,1)]);
  b = a[1] + a[2]^2;
end ConstRepl;
```

**Comment by casella** [comment]
> I just tested with the current master (#11507 included) and `s` is computed efficiently as e.g. `s[1,1] = 0.6666666666666666`.

Great, thanks! I'll check with the nightly build if we can actually manage the flexible multibody system model.

**Comment by casella** [comment]
> With the new backend the `cat` expression is not simplified since we don't expand `fill` calls, so it ends up as `(cat(2, fill(0, 6, 3 * (i - 2)), identity(6), fill(0, 6, 3 * (3 - i))));` in the flat model. The new backend probably simplifies this at some point, but I'm not sure how to check that.

I guess how to do that best is still an open problem. On one hand, one would like to avoid expanding arrays, on the other hand handling such generic expressions efficiently could be non-trival. I'm not really sure what is the best way to proceed.

For sure the MWE of this ticket is a nice test case with practical applications (FEM multibody systems).
> Edit: Just to be clear, this only fixes the issue with `B[2, :, :]`. The binding of `s` still ends up as a huge expression, but I'm not sure what can be done about that.

**Comment by casella** [comment]
I finally had time to test `Test3DArray` after @perost's pull request.

The good news is that the backend now handles all the simplifications efficiently: I scanned the solved model with the declarative debugger and indeed each element of `B`, `Sbar`, and `s` is explicitly assigned to zero or to a constant literal value. So, it seems that now that cat() is successfully removed by the frontend, the backend does its job nicely.

The bad news is that the code generation still takes forever. In the N = 3 case, on my PC it takes 1 s to go through all the steps until SimCode, but then Templates takes two minutes (?!?). I guess the reason is related to the generation of the `_info.json` file, which is a whooping 183 MB for a model that has 243 variables. I'll open a separate ticket for that.

**Comment by casella** [comment]
In fact, it's the logging of the simplification operations 😄. Some variables have over 1 MB of log 😨

**Comment by casella** [comment]
If I disable "Generate operations", the model compiles and runs in no time.

**Comment by casella** [comment]
@perost sorry but I need to reopen this. I tried to run the original model, which is the Flex.Test.FreeVibration model in https://github.com/casella/FlexBeam, and it just takes forever to pass through the NF.

Before we delve into that, one easy question specific to this ticket. After #11507, the element of `B[:,:,:]` that are zero and one are evaluated correctly as such. However, the equation for s is the following:
```modelica
s = {{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0}} + {{(B[1,1,1] * 0.3333333333333333 + B[1,4,1] * 0.16666666666666666) * B[1,1,1] + (B[1,1,1] * 0.16666666666666666 + B[1,4,1] * 0.3333333333333333) * B[1,4,1], (B[1,1,1] * 0.3333333333333333 + B[1,4,1] * 0.16666666666666666) * B[1,1,2] + (B[1,1,1] * 0.16666666666666666 + B[1,4,1] * 0.3333333333333333) * B[1,4,2], (B[1,1,1] * 0.3333333333333333 + B[1,4,1] * 0.16666666666666666) * B[1,1,3] + (B[1,1,1] * 0.16666666666666666 + B[1,4,1] * 0.3333333333333333) * B[1,4,3], (B[1,1,1] * 0.3333333333333333 + B[1,4,1] * 0.16666666666666666) * B[1,1,4] + (B[1,1,1] * 0.16666666666666666 + B[1,4,1] * 0.3333333333333333) * B[1,4,4], (B[1,1,1] * 0.3333333333333333 + B[1,4,1] * 0.16666666666666666) * B[1,1,5] + (B[1,1,1] * 0.16666666666666666 + B[1,4,1] * 0.3333333333333333) * B[1,4,5], (B[1,1,1] * 0.3333333333333333 + B[1,4,1] * 0.16666666666666666) * B[1,1,6] + (B[1,1,1] * 0.16666666666666666 + B[1,4,1] * 0.3333333333333333) * B[1,4,6], (B[1,1,1] * 0.3333333333333333 + B[1,4,1] * 0.16666666666666666) * B[1,1,7] + (B[1,1,1] * 0.16666666666666666 + B[1,4,1] * 0.3333333333333333) * B[1,4,7], (B[1,1,1] * 0.3333333333333333 + B[1,4,1] * 0.16666666666666666) * B[1,1,8] + (B[1,1,1] * 0.16666666666666666 + B[1,4,1] * 0.3333333333333333) * B[1,4,8], (B[1,1,1] * 0.3333333333333333 + B[1,4,1] * 0.16666666666666666) * B[1,1,9] + (B[1,1,1] * 0.16666666666666666 + B[1,4,1] * 0.3333333333333333) * B[1,4,9]}, {(B[1,1,2] * 0.3333333333333333 + B[1,4,2] * 0.16666666666666666) * B[1,1,1] + (B[1,1,2] * 0.16666666666666666 + B[1,4,2] * 0.3333333333333333) * B[1,4,1], (B[1,1,2] * 0.3333333333333333 + B[1,4,2] * 0.16666666666666666) * B[1,1,2] + (B[1,1,2] * 0.16666666666666666 + B[1,4,2] * 0.3333333333333333) * B[1,4,2], (B[1,1,2] * 0.3333333333333333 + B[1,4,2] * 0.16666666666666666) * B[1,1,3] + (B[1,1,2] * 0.16666666666666666 + B[1,4,2] * 0.3333333333333333) * B[1,4,3], (B[1,1,2] * 0.3333333333333333 + B[1,4,2] * 0.16666666666666666) * B[1,1,4] + (B[1,1,2] * 0.16666666666666666 + B[1,4,2] * 0.3333333333333333) * B[1,4,4], (B[1,1,2] * 0.3333333333333333 + B[1,4,2] * 0.16666666666666666) * B[1,1,5] + (B[1,1,2] * 0.16666666666666666 + B[1,4,2] * 0.3333333333333333) * B[1,4,5], (B[1,1,2] * 0.3333333333333333 + B[1,4,2] * 0.16666666666666666) * B[1,1,6] + (B[1,1,2] * 0.16666666666666666 + B[1,4,2] * 0.3333333333333333) * B[1,4,6], (B[1,1,2] * 0.3333333333333333 + B[1,4,2] * 0.16666666666666666) * B[1,1,7] + (B[1,1,2] * 0.16666666666666666 + B[1,4,2] * 0.3333333333333333) * B[1,4,7], (B[1,1,2] * 0.3333333333333333 + B[1,4,2] * 0.16666666666666666) * B[1,1,8] + (B[1,1,2] * 0.16666666666666666 + B[1,4,2] * 0.3333333333333333) * B[1,4,8], (B[1,1,2] * 0.3333333333333333 + B[1,4,2] * 0.16666666666666666) * B[1,1,9] + (B[1,1,2] * 0.16666666666666666 + B[1,4,2] * 0.3333333333333333) * B[1,4,9]}, {(B[1,1,3] * 0.3333333333333333 + B[1,4,3] * 0.16666666666666666) * B[1,1,1] + (B[1,1,3] * 0.16666666666666666 + B[1,4,3] * 0.3333333333333333) * B[1,4,1], (B[1,1,3] * 0.3333333333333333 + B[1,4,3] * 0.16666666666666666) * B[1,1,2] + (B[1,1,3] * 0.16666666666666666 + B[1,4,3] * 0.3333333333333333) * B[1,4,2], (B[1,1,3] * 0.3333333333333333 + B[1,4,3] * 0.16666666666666666) * B[1,1,3] + (B[1,1,3] * 0.16666666666666666 + B[1,4,3] * 0.3333333333333333) * B[1,4,3], (B[1,1,3] * 0.3333333333333333 + B[1,4,3] * 0.16666666666666666) * B[1,1,4] + (B[1,1,3] * 0.16666666666666666 + B[1,4,3] * 0.3333333333333333) * B[1,4,4], (B[1,1,3] * 0.3333333333333333 + B[1,4,3] * 0.16666666666666666) * B[1,1,5] + (B[1,1,3] * 0.16666666666666666 + B[1,4,3] * 0.3333333333333333) * B[1,4,5], (B[1,1,3] * 0.3333333333333333 + B[1,4,3] * 0.16666666666666666) * B[1,1,6] + (B[1,1,3] * 0.16666666666666666 + B[1,4,3] * 0.3333333333333333) * B[1,4,6], (B[1,1,3] * 0.3333333333333333 + B[1,4,3] * 0.16666666666666666) * B[1,1,7] + (B[1,1,3] * 0.16666666666666666 + B[1,4,3] * 0.3333333333333333) * B[1,4,7], (B[1,1,3] * 0.3333333333333333 + B[1,4,3] * 0.16666666666666666) * B[1,1,8] + (B[1,1,3] * 0.16666666666666666 + B[1,4,3] * 0.3333333333333333) * B[1,4,8], (B[1,1,3] * 0.3333333333333333 + B[1,4,3] * 0.16666666666666666) * B[1,1,9] + (B[1,1,3] * 0.16666666666666666 + B[1,4,3] * 0.3333333333333333) * B[1,4,9]}, {(B[1,1,4] * 0.3333333333333333 + B[1,4,4] * 0.16666666666666666) * B[1,1,1] + (B[1,1,4] * 0.16666666666666666 + B[1,4,4] * 0.3333333333333333) * B[1,4,1], (B[1,1,4] * 0.3333333333333333 + B[1,4,4] * 0.16666666666666666) * B[1,1,2] + (B[1,1,4] * 0.16666666666666666 + B[1,4,4] * 0.3333333333333333) * B[1,4,2], (B[1,1,4] * 0.3333333333333333 + B[1,4,4] * 0.16666666666666666) * B[1,1,3] + (B[1,1,4] * 0.16666666666666666 + B[1,4,4] * 0.3333333333333333) * B[1,4,3], (B[1,1,4] * 0.3333333333333333 + B[1,4,4] * 0.16666666666666666) * B[1,1,4] + (B[1,1,4] * 0.16666666666666666 + B[1,4,4] * 0.3333333333333333) * B[1,4,4], (B[1,1,4] * 0.3333333333333333 + B[1,4,4] * 0.16666666666666666) * B[1,1,5] + (B[1,1,4] * 0.16666666666666666 + B[1,4,4] * 0.3333333333333333) * B[1,4,5], (B[1,1,4] * 0.3333333333333333 + B[1,4,4] * 0.16666666666666666) * B[1,1,6] + (B[1,1,4] * 0.16666666666666666 + B[1,4,4] * 0.3333333333333333) * B[1,4,6], (B[1,1,4] * 0.3333333333333333 + B[1,4,4] * 0.16666666666666666) * B[1,1,7] + (B[1,1,4] * 0.16666666666666666 + B[1,4,4] * 0.3333333333333333) * B[1,4,7], (B[1,1,4] * 0.3333333333333333 + B[1,4,4] * 0.16666666666666666) * B[1,1,8] + (B[1,1,4] * 0.16666666666666666 + B[1,4,4] * 0.3333333333333333) * B[1,4,8], (B[1,1,4] * 0.3333333333333333 + B[1,4,4] * 0.16666666666666666) * B[1,1,9] + (B[1,1,4] * 0.16666666666666666 + B[1,4,4] * 0.3333333333333333) * B[1,4,9]}, {(B[1,1,5] * 0.3333333333333333 + B[1,4,5] * 0.16666666666666666) * B[1,1,1] + (B[1,1,5] * 0.16666666666666666 + B[1,4,5] * 0.3333333333333333) * B[1,4,1], (B[1,1,5] * 0.3333333333333333 + B[1,4,5] * 0.16666666666666666) * B[1,1,2] + (B[1,1,5] * 0.16666666666666666 + B[1,4,5] * 0.3333333333333333) * B[1,4,2], (B[1,1,5] * 0.3333333333333333 + B[1,4,5] * 0.16666666666666666) * B[1,1,3] + (B[1,1,5] * 0.16666666666666666 + B[1,4,5] * 0.3333333333333333) * B[1,4,3], (B[1,1,5] * 0.3333333333333333 + B[1,4,5] * 0.16666666666666666) * B[1,1,4] + (B[1,1,5] * 0.16666666666666666 + B[1,4,5] * 0.3333333333333333) * B[1,4,4], (B[1,1,5] * 0.3333333333333333 + B[1,4,5] * 0.16666666666666666) * B[1,1,5] + (B[1,1,5] * 0.16666666666666666 + B[1,4,5] * 0.3333333333333333) * B[1,4,5], (B[1,1,5] * 0.3333333333333333 + B[1,4,5] * 0.16666666666666666) * B[1,1,6] + (B[1,1,5] * 0.16666666666666666 + B[1,4,5] * 0.3333333333333333) * B[1,4,6], (B[1,1,5] * 0.3333333333333333 + B[1,4,5] * 0.16666666666666666) * B[1,1,7] + (B[1,1,5] * 0.16666666666666666 + B[1,4,5] * 0.3333333333333333) * B[1,4,7], (B[1,1,5] * 0.3333333333333333 + B[1,4,5] * 0.16666666666666666) * B[1,1,8] + (B[1,1,5] * 0.16666666666666666 + B[1,4,5] * 0.3333333333333333) * B[1,4,8], (B[1,1,5] * 0.3333333333333333 + B[1,4,5] * 0.16666666666666666) * B[1,1,9] + (B[1,1,5] * 0.16666666666666666 + B[1,4,5] * 0.3333333333333333) * B[1,4,9]}, {(B[1,1,6] * 0.3333333333333333 + B[1,4,6] * 0.16666666666666666) * B[1,1,1] + (B[1,1,6] * 0.16666666666666666 + B[1,4,6] * 0.3333333333333333) * B[1,4,1], (B[1,1,6] * 0.3333333333333333 + B[1,4,6] * 0.16666666666666666) * B[1,1,2] + (B[1,1,6] * 0.16666666666666666 + B[1,4,6] * 0.3333333333333333) * B[1,4,2], (B[1,1,6] * 0.3333333333333333 + B[1,4,6] * 0.16666666666666666) * B[1,1,3] + (B[1,1,6] * 0.16666666666666666 + B[1,4,6] * 0.3333333333333333) * B[1,4,3], (B[1,1,6] * 0.3333333333333333 + B[1,4,6] * 0.16666666666666666) * B[1,1,4] + (B[1,1,6] * 0.16666666666666666 + B[1,4,6] * 0.3333333333333333) * B[1,4,4], (B[1,1,6] * 0.3333333333333333 + B[1,4,6] * 0.16666666666666666) * B[1,1,5] + (B[1,1,6] * 0.16666666666666666 + B[1,4,6] * 0.3333333333333333) * B[1,4,5], (B[1,1,6] * 0.3333333333333333 + B[1,4,6] * 0.16666666666666666) * B[1,1,6] + (B[1,1,6] * 0.16666666666666666 + B[1,4,6] * 0.3333333333333333) * B[1,4,6], (B[1,1,6] * 0.3333333333333333 + B[1,4,6] * 0.16666666666666666) * B[1,1,7] + (B[1,1,6] * 0.16666666666666666 + B[1,4,6] * 0.3333333333333333) * B[1,4,7], (B[1,1,6] * 0.3333333333333333 + B[1,4,6] * 0.16666666666666666) * B[1,1,8] + (B[1,1,6] * 0.16666666666666666 + B[1,4,6] * 0.3333333333333333) * B[1,4,8], (B[1,1,6] * 0.3333333333333333 + B[1,4,6] * 0.16666666666666666) * B[1,1,9] + (B[1,1,6] * 0.16666666666666666 + B[1,4,6] * 0.3333333333333333) * B[1,4,9]}, {(B[1,1,7] * 0.3333333333333333 + B[1,4,7] * 0.16666666666666666) * B[1,1,1] + (B[1,1,7] * 0.16666666666666666 + B[1,4,7] * 0.3333333333333333) * B[1,4,1], (B[1,1,7] * 0.3333333333333333 + B[1,4,7] * 0.16666666666666666) * B[1,1,2] + (B[1,1,7] * 0.16666666666666666 + B[1,4,7] * 0.3333333333333333) * B[1,4,2], (B[1,1,7] * 0.3333333333333333 + B[1,4,7] * 0.16666666666666666) * B[1,1,3] + (B[1,1,7] * 0.16666666666666666 + B[1,4,7] * 0.3333333333333333) * B[1,4,3], (B[1,1,7] * 0.3333333333333333 + B[1,4,7] * 0.16666666666666666) * B[1,1,4] + (B[1,1,7] * 0.16666666666666666 + B[1,4,7] * 0.3333333333333333) * B[1,4,4], (B[1,1,7] * 0.3333333333333333 + B[1,4,7] * 0.16666666666666666) * B[1,1,5] + (B[1,1,7] * 0.16666666666666666 + B[1,4,7] * 0.3333333333333333) * B[1,4,5], (B[1,1,7] * 0.3333333333333333 + B[1,4,7] * 0.16666666666666666) * B[1,1,6] + (B[1,1,7] * 0.16666666666666666 + B[1,4,7] * 0.3333333333333333) * B[1,4,6], (B[1,1,7] * 0.3333333333333333 + B[1,4,7] * 0.16666666666666666) * B[1,1,7] + (B[1,1,7] * 0.16666666666666666 + B[1,4,7] * 0.3333333333333333) * B[1,4,7], (B[1,1,7] * 0.3333333333333333 + B[1,4,7] * 0.16666666666666666) * B[1,1,8] + (B[1,1,7] * 0.16666666666666666 + B[1,4,7] * 0.3333333333333333) * B[1,4,8], (B[1,1,7] * 0.3333333333333333 + B[1,4,7] * 0.16666666666666666) * B[1,1,9] + (B[1,1,7] * 0.16666666666666666 + B[1,4,7] * 0.3333333333333333) * B[1,4,9]}, {(B[1,1,8] * 0.3333333333333333 + B[1,4,8] * 0.16666666666666666) * B[1,1,1] + (B[1,1,8] * 0.16666666666666666 + B[1,4,8] * 0.3333333333333333) * B[1,4,1], (B[1,1,8] * 0.3333333333333333 + B[1,4,8] * 0.16666666666666666) * B[1,1,2] + (B[1,1,8] * 0.16666666666666666 + B[1,4,8] * 0.3333333333333333) * B[1,4,2], (B[1,1,8] * 0.3333333333333333 + B[1,4,8] * 0.16666666666666666) * B[1,1,3] + (B[1,1,8] * 0.16666666666666666 + B[1,4,8] * 0.3333333333333333) * B[1,4,3], (B[1,1,8] * 0.3333333333333333 + B[1,4,8] * 0.16666666666666666) * B[1,1,4] + (B[1,1,8] * 0.16666666666666666 + B[1,4,8] * 0.3333333333333333) * B[1,4,4], (B[1,1,8] * 0.3333333333333333 + B[1,4,8] * 0.16666666666666666) * B[1,1,5] + (B[1,1,8] * 0.16666666666666666 + B[1,4,8] * 0.3333333333333333) * B[1,4,5], (B[1,1,8] * 0.3333333333333333 + B[1,4,8] * 0.16666666666666666) * B[1,1,6] + (B[1,1,8] * 0.16666666666666666 + B[1,4,8] * 0.3333333333333333) * B[1,4,6], (B[1,1,8] * 0.3333333333333333 + B[1,4,8] * 0.16666666666666666) * B[1,1,7] + (B[1,1,8] * 0.16666666666666666 + B[1,4,8] * 0.3333333333333333) * B[1,4,7], (B[1,1,8] * 0.3333333333333333 + B[1,4,8] * 0.16666666666666666) * B[1,1,8] + (B[1,1,8] * 0.16666666666666666 + B[1,4,8] * 0.3333333333333333) * B[1,4,8], (B[1,1,8] * 0.3333333333333333 + B[1,4,8] * 0.16666666666666666) * B[1,1,9] + (B[1,1,8] * 0.16666666666666666 + B[1,4,8] * 0.3333333333333333) * B[1,4,9]}, {(B[1,1,9] * 0.3333333333333333 + B[1,4,9] * 0.16666666666666666) * B[1,1,1] + (B[1,1,9] * 0.16666666666666666 + B[1,4,9] * 0.3333333333333333) * B[1,4,1], (B[1,1,9] * 0.3333333333333333 + B[1,4,9] * 0.16666666666666666) * B[1,1,2] + (B[1,1,9] * 0.16666666666666666 + B[1,4,9] * 0.3333333333333333) * B[1,4,2], (B[1,1,9] * 0.3333333333333333 + B[1,4,9] * 0.16666666666666666) * B[1,1,3] + (B[1,1,9] * 0.16666666666666666 + B[1,4,9] * 0.3333333333333333) * B[1,4,3], (B[1,1,9] * 0.3333333333333333 + B[1,4,9] * 0.16666666666666666) * B[1,1,4] + (B[1,1,9] * 0.16666666666666666 + B[1,4,9] * 0.3333333333333333) * B[1,4,4], (B[1,1,9] * 0.3333333333333333 + B[1,4,9] * 0.16666666666666666) * B[1,1,5] + (B[1,1,9] * 0.16666666666666666 + B[1,4,9] * 0.3333333333333333) * B[1,4,5], (B[1,1,9] * 0.3333333333333333 + B[1,4,9] * 0.16666666666666666) * B[1,1,6] + (B[1,1,9] * 0.16666666666666666 + B[1,4,9] * 0.3333333333333333) * B[1,4,6], (B[1,1,9] * 0.3333333333333333 + B[1,4,9] * 0.16666666666666666) * B[1,1,7] + (B[1,1,9] * 0.16666666666666666 + B[1,4,9] * 0.3333333333333333) * B[1,4,7], (B[1,1,9] * 0.3333333333333333 + B[1,4,9] * 0.16666666666666666) * B[1,1,8] + (B[1,1,9] * 0.16666666666666666 + B[1,4,9] * 0.3333333333333333) * B[1,4,8], (B[1,1,9] * 0.3333333333333333 + B[1,4,9] * 0.16666666666666666) * B[1,1,9] + (B[1,1,9] * 0.16666666666666666 + B[1,4,9] * 0.3333333333333333) * B[1,4,9]}} + {{(B[2,1,1] * 0.3333333333333333 + B[2,4,1] * 0.16666666666666666) * B[2,1,1] + (B[2,1,1] * 0.16666666666666666 + B[2,4,1] * 0.3333333333333333) * B[2,4,1], (B[2,1,1] * 0.3333333333333333 + B[2,4,1] * 0.16666666666666666) * B[2,1,2] + (B[2,1,1] * 0.16666666666666666 + B[2,4,1] * 0.3333333333333333) * B[2,4,2], (B[2,1,1] * 0.3333333333333333 + B[2,4,1] * 0.16666666666666666) * B[2,1,3] + (B[2,1,1] * 0.16666666666666666 + B[2,4,1] * 0.3333333333333333) * B[2,4,3], (B[2,1,1] * 0.3333333333333333 + B[2,4,1] * 0.16666666666666666) * B[2,1,4] + (B[2,1,1] * 0.16666666666666666 + B[2,4,1] * 0.3333333333333333) * B[2,4,4], (B[2,1,1] * 0.3333333333333333 + B[2,4,1] * 0.16666666666666666) * B[2,1,5] + (B[2,1,1] * 0.16666666666666666 + B[2,4,1] * 0.3333333333333333) * B[2,4,5], (B[2,1,1] * 0.3333333333333333 + B[2,4,1] * 0.16666666666666666) * B[2,1,6] + (B[2,1,1] * 0.16666666666666666 + B[2,4,1] * 0.3333333333333333) * B[2,4,6], (B[2,1,1] * 0.3333333333333333 + B[2,4,1] * 0.16666666666666666) * B[2,1,7] + (B[2,1,1] * 0.16666666666666666 + B[2,4,1] * 0.3333333333333333) * B[2,4,7], (B[2,1,1] * 0.3333333333333333 + B[2,4,1] * 0.16666666666666666) * B[2,1,8] + (B[2,1,1] * 0.16666666666666666 + B[2,4,1] * 0.3333333333333333) * B[2,4,8], (B[2,1,1] * 0.3333333333333333 + B[2,4,1] * 0.16666666666666666) * B[2,1,9] + (B[2,1,1] * 0.16666666666666666 + B[2,4,1] * 0.3333333333333333) * B[2,4,9]}, {(B[2,1,2] * 0.3333333333333333 + B[2,4,2] * 0.16666666666666666) * B[2,1,1] + (B[2,1,2] * 0.16666666666666666 + B[2,4,2] * 0.3333333333333333) * B[2,4,1], (B[2,1,2] * 0.3333333333333333 + B[2,4,2] * 0.16666666666666666) * B[2,1,2] + (B[2,1,2] * 0.16666666666666666 + B[2,4,2] * 0.3333333333333333) * B[2,4,2], (B[2,1,2] * 0.3333333333333333 + B[2,4,2] * 0.16666666666666666) * B[2,1,3] + (B[2,1,2] * 0.16666666666666666 + B[2,4,2] * 0.3333333333333333) * B[2,4,3], (B[2,1,2] * 0.3333333333333333 + B[2,4,2] * 0.16666666666666666) * B[2,1,4] + (B[2,1,2] * 0.16666666666666666 + B[2,4,2] * 0.3333333333333333) * B[2,4,4], (B[2,1,2] * 0.3333333333333333 + B[2,4,2] * 0.16666666666666666) * B[2,1,5] + (B[2,1,2] * 0.16666666666666666 + B[2,4,2] * 0.3333333333333333) * B[2,4,5], (B[2,1,2] * 0.3333333333333333 + B[2,4,2] * 0.16666666666666666) * B[2,1,6] + (B[2,1,2] * 0.16666666666666666 + B[2,4,2] * 0.3333333333333333) * B[2,4,6], (B[2,1,2] * 0.3333333333333333 + B[2,4,2] * 0.16666666666666666) * B[2,1,7] + (B[2,1,2] * 0.16666666666666666 + B[2,4,2] * 0.3333333333333333) * B[2,4,7], (B[2,1,2] * 0.3333333333333333 + B[2,4,2] * 0.16666666666666666) * B[2,1,8] + (B[2,1,2] * 0.16666666666666666 + B[2,4,2] * 0.3333333333333333) * B[2,4,8], (B[2,1,2] * 0.3333333333333333 + B[2,4,2] * 0.16666666666666666) * B[2,1,9] + (B[2,1,2] * 0.16666666666666666 + B[2,4,2] * 0.3333333333333333) * B[2,4,9]}, {(B[2,1,3] * 0.3333333333333333 + B[2,4,3] * 0.16666666666666666) * B[2,1,1] + (B[2,1,3] * 0.16666666666666666 + B[2,4,3] * 0.3333333333333333) * B[2,4,1], (B[2,1,3] * 0.3333333333333333 + B[2,4,3] * 0.16666666666666666) * B[2,1,2] + (B[2,1,3] * 0.16666666666666666 + B[2,4,3] * 0.3333333333333333) * B[2,4,2], (B[2,1,3] * 0.3333333333333333 + B[2,4,3] * 0.16666666666666666) * B[2,1,3] + (B[2,1,3] * 0.16666666666666666 + B[2,4,3] * 0.3333333333333333) * B[2,4,3], (B[2,1,3] * 0.3333333333333333 + B[2,4,3] * 0.16666666666666666) * B[2,1,4] + (B[2,1,3] * 0.16666666666666666 + B[2,4,3] * 0.3333333333333333) * B[2,4,4], (B[2,1,3] * 0.3333333333333333 + B[2,4,3] * 0.16666666666666666) * B[2,1,5] + (B[2,1,3] * 0.16666666666666666 + B[2,4,3] * 0.3333333333333333) * B[2,4,5], (B[2,1,3] * 0.3333333333333333 + B[2,4,3] * 0.16666666666666666) * B[2,1,6] + (B[2,1,3] * 0.16666666666666666 + B[2,4,3] * 0.3333333333333333) * B[2,4,6], (B[2,1,3] * 0.3333333333333333 + B[2,4,3] * 0.16666666666666666) * B[2,1,7] + (B[2,1,3] * 0.16666666666666666 + B[2,4,3] * 0.3333333333333333) * B[2,4,7], (B[2,1,3] * 0.3333333333333333 + B[2,4,3] * 0.16666666666666666) * B[2,1,8] + (B[2,1,3] * 0.16666666666666666 + B[2,4,3] * 0.3333333333333333) * B[2,4,8], (B[2,1,3] * 0.3333333333333333 + B[2,4,3] * 0.16666666666666666) * B[2,1,9] + (B[2,1,3] * 0.16666666666666666 + B[2,4,3] * 0.3333333333333333) * B[2,4,9]}, {(B[2,1,4] * 0.3333333333333333 + B[2,4,4] * 0.16666666666666666) * B[2,1,1] + (B[2,1,4] * 0.16666666666666666 + B[2,4,4] * 0.3333333333333333) * B[2,4,1], (B[2,1,4] * 0.3333333333333333 + B[2,4,4] * 0.16666666666666666) * B[2,1,2] + (B[2,1,4] * 0.16666666666666666 + B[2,4,4] * 0.3333333333333333) * B[2,4,2], (B[2,1,4] * 0.3333333333333333 + B[2,4,4] * 0.16666666666666666) * B[2,1,3] + (B[2,1,4] * 0.16666666666666666 + B[2,4,4] * 0.3333333333333333) * B[2,4,3], (B[2,1,4] * 0.3333333333333333 + B[2,4,4] * 0.16666666666666666) * B[2,1,4] + (B[2,1,4] * 0.16666666666666666 + B[2,4,4] * 0.3333333333333333) * B[2,4,4], (B[2,1,4] * 0.3333333333333333 + B[2,4,4] * 0.16666666666666666) * B[2,1,5] + (B[2,1,4] * 0.16666666666666666 + B[2,4,4] * 0.3333333333333333) * B[2,4,5], (B[2,1,4] * 0.3333333333333333 + B[2,4,4] * 0.16666666666666666) * B[2,1,6] + (B[2,1,4] * 0.16666666666666666 + B[2,4,4] * 0.3333333333333333) * B[2,4,6], (B[2,1,4] * 0.3333333333333333 + B[2,4,4] * 0.16666666666666666) * B[2,1,7] + (B[2,1,4] * 0.16666666666666666 + B[2,4,4] * 0.3333333333333333) * B[2,4,7], (B[2,1,4] * 0.3333333333333333 + B[2,4,4] * 0.16666666666666666) * B[2,1,8] + (B[2,1,4] * 0.16666666666666666 + B[2,4,4] * 0.3333333333333333) * B[2,4,8], (B[2,1,4] * 0.3333333333333333 + B[2,4,4] * 0.16666666666666666) * B[2,1,9] + (B[2,1,4] * 0.16666666666666666 + B[2,4,4] * 0.3333333333333333) * B[2,4,9]}, {(B[2,1,5] * 0.3333333333333333 + B[2,4,5] * 0.16666666666666666) * B[2,1,1] + (B[2,1,5] * 0.16666666666666666 + B[2,4,5] * 0.3333333333333333) * B[2,4,1], (B[2,1,5] * 0.3333333333333333 + B[2,4,5] * 0.16666666666666666) * B[2,1,2] + (B[2,1,5] * 0.16666666666666666 + B[2,4,5] * 0.3333333333333333) * B[2,4,2], (B[2,1,5] * 0.3333333333333333 + B[2,4,5] * 0.16666666666666666) * B[2,1,3] + (B[2,1,5] * 0.16666666666666666 + B[2,4,5] * 0.3333333333333333) * B[2,4,3], (B[2,1,5] * 0.3333333333333333 + B[2,4,5] * 0.16666666666666666) * B[2,1,4] + (B[2,1,5] * 0.16666666666666666 + B[2,4,5] * 0.3333333333333333) * B[2,4,4], (B[2,1,5] * 0.3333333333333333 + B[2,4,5] * 0.16666666666666666) * B[2,1,5] + (B[2,1,5] * 0.16666666666666666 + B[2,4,5] * 0.3333333333333333) * B[2,4,5], (B[2,1,5] * 0.3333333333333333 + B[2,4,5] * 0.16666666666666666) * B[2,1,6] + (B[2,1,5] * 0.16666666666666666 + B[2,4,5] * 0.3333333333333333) * B[2,4,6], (B[2,1,5] * 0.3333333333333333 + B[2,4,5] * 0.16666666666666666) * B[2,1,7] + (B[2,1,5] * 0.16666666666666666 + B[2,4,5] * 0.3333333333333333) * B[2,4,7], (B[2,1,5] * 0.3333333333333333 + B[2,4,5] * 0.16666666666666666) * B[2,1,8] + (B[2,1,5] * 0.16666666666666666 + B[2,4,5] * 0.3333333333333333) * B[2,4,8], (B[2,1,5] * 0.3333333333333333 + B[2,4,5] * 0.16666666666666666) * B[2,1,9] + (B[2,1,5] * 0.16666666666666666 + B[2,4,5] * 0.3333333333333333) * B[2,4,9]}, {(B[2,1,6] * 0.3333333333333333 + B[2,4,6] * 0.16666666666666666) * B[2,1,1] + (B[2,1,6] * 0.16666666666666666 + B[2,4,6] * 0.3333333333333333) * B[2,4,1], (B[2,1,6] * 0.3333333333333333 + B[2,4,6] * 0.16666666666666666) * B[2,1,2] + (B[2,1,6] * 0.16666666666666666 + B[2,4,6] * 0.3333333333333333) * B[2,4,2], (B[2,1,6] * 0.3333333333333333 + B[2,4,6] * 0.16666666666666666) * B[2,1,3] + (B[2,1,6] * 0.16666666666666666 + B[2,4,6] * 0.3333333333333333) * B[2,4,3], (B[2,1,6] * 0.3333333333333333 + B[2,4,6] * 0.16666666666666666) * B[2,1,4] + (B[2,1,6] * 0.16666666666666666 + B[2,4,6] * 0.3333333333333333) * B[2,4,4], (B[2,1,6] * 0.3333333333333333 + B[2,4,6] * 0.16666666666666666) * B[2,1,5] + (B[2,1,6] * 0.16666666666666666 + B[2,4,6] * 0.3333333333333333) * B[2,4,5], (B[2,1,6] * 0.3333333333333333 + B[2,4,6] * 0.16666666666666666) * B[2,1,6] + (B[2,1,6] * 0.16666666666666666 + B[2,4,6] * 0.3333333333333333) * B[2,4,6], (B[2,1,6] * 0.3333333333333333 + B[2,4,6] * 0.16666666666666666) * B[2,1,7] + (B[2,1,6] * 0.16666666666666666 + B[2,4,6] * 0.3333333333333333) * B[2,4,7], (B[2,1,6] * 0.3333333333333333 + B[2,4,6] * 0.16666666666666666) * B[2,1,8] + (B[2,1,6] * 0.16666666666666666 + B[2,4,6] * 0.3333333333333333) * B[2,4,8], (B[2,1,6] * 0.3333333333333333 + B[2,4,6] * 0.16666666666666666) * B[2,1,9] + (B[2,1,6] * 0.16666666666666666 + B[2,4,6] * 0.3333333333333333) * B[2,4,9]}, {(B[2,1,7] * 0.3333333333333333 + B[2,4,7] * 0.16666666666666666) * B[2,1,1] + (B[2,1,7] * 0.16666666666666666 + B[2,4,7] * 0.3333333333333333) * B[2,4,1], (B[2,1,7] * 0.3333333333333333 + B[2,4,7] * 0.16666666666666666) * B[2,1,2] + (B[2,1,7] * 0.16666666666666666 + B[2,4,7] * 0.3333333333333333) * B[2,4,2], (B[2,1,7] * 0.3333333333333333 + B[2,4,7] * 0.16666666666666666) * B[2,1,3] + (B[2,1,7] * 0.16666666666666666 + B[2,4,7] * 0.3333333333333333) * B[2,4,3], (B[2,1,7] * 0.3333333333333333 + B[2,4,7] * 0.16666666666666666) * B[2,1,4] + (B[2,1,7] * 0.16666666666666666 + B[2,4,7] * 0.3333333333333333) * B[2,4,4], (B[2,1,7] * 0.3333333333333333 + B[2,4,7] * 0.16666666666666666) * B[2,1,5] + (B[2,1,7] * 0.16666666666666666 + B[2,4,7] * 0.3333333333333333) * B[2,4,5], (B[2,1,7] * 0.3333333333333333 + B[2,4,7] * 0.16666666666666666) * B[2,1,6] + (B[2,1,7] * 0.16666666666666666 + B[2,4,7] * 0.3333333333333333) * B[2,4,6], (B[2,1,7] * 0.3333333333333333 + B[2,4,7] * 0.16666666666666666) * B[2,1,7] + (B[2,1,7] * 0.16666666666666666 + B[2,4,7] * 0.3333333333333333) * B[2,4,7], (B[2,1,7] * 0.3333333333333333 + B[2,4,7] * 0.16666666666666666) * B[2,1,8] + (B[2,1,7] * 0.16666666666666666 + B[2,4,7] * 0.3333333333333333) * B[2,4,8], (B[2,1,7] * 0.3333333333333333 + B[2,4,7] * 0.16666666666666666) * B[2,1,9] + (B[2,1,7] * 0.16666666666666666 + B[2,4,7] * 0.3333333333333333) * B[2,4,9]}, {(B[2,1,8] * 0.3333333333333333 + B[2,4,8] * 0.16666666666666666) * B[2,1,1] + (B[2,1,8] * 0.16666666666666666 + B[2,4,8] * 0.3333333333333333) * B[2,4,1], (B[2,1,8] * 0.3333333333333333 + B[2,4,8] * 0.16666666666666666) * B[2,1,2] + (B[2,1,8] * 0.16666666666666666 + B[2,4,8] * 0.3333333333333333) * B[2,4,2], (B[2,1,8] * 0.3333333333333333 + B[2,4,8] * 0.16666666666666666) * B[2,1,3] + (B[2,1,8] * 0.16666666666666666 + B[2,4,8] * 0.3333333333333333) * B[2,4,3], (B[2,1,8] * 0.3333333333333333 + B[2,4,8] * 0.16666666666666666) * B[2,1,4] + (B[2,1,8] * 0.16666666666666666 + B[2,4,8] * 0.3333333333333333) * B[2,4,4], (B[2,1,8] * 0.3333333333333333 + B[2,4,8] * 0.16666666666666666) * B[2,1,5] + (B[2,1,8] * 0.16666666666666666 + B[2,4,8] * 0.3333333333333333) * B[2,4,5], (B[2,1,8] * 0.3333333333333333 + B[2,4,8] * 0.16666666666666666) * B[2,1,6] + (B[2,1,8] * 0.16666666666666666 + B[2,4,8] * 0.3333333333333333) * B[2,4,6], (B[2,1,8] * 0.3333333333333333 + B[2,4,8] * 0.16666666666666666) * B[2,1,7] + (B[2,1,8] * 0.16666666666666666 + B[2,4,8] * 0.3333333333333333) * B[2,4,7], (B[2,1,8] * 0.3333333333333333 + B[2,4,8] * 0.16666666666666666) * B[2,1,8] + (B[2,1,8] * 0.16666666666666666 + B[2,4,8] * 0.3333333333333333) * B[2,4,8], (B[2,1,8] * 0.3333333333333333 + B[2,4,8] * 0.16666666666666666) * B[2,1,9] + (B[2,1,8] * 0.16666666666666666 + B[2,4,8] * 0.3333333333333333) * B[2,4,9]}, {(B[2,1,9] * 0.3333333333333333 + B[2,4,9] * 0.16666666666666666) * B[2,1,1] + (B[2,1,9] * 0.16666666666666666 + B[2,4,9] * 0.3333333333333333) * B[2,4,1], (B[2,1,9] * 0.3333333333333333 + B[2,4,9] * 0.16666666666666666) * B[2,1,2] + (B[2,1,9] * 0.16666666666666666 + B[2,4,9] * 0.3333333333333333) * B[2,4,2], (B[2,1,9] * 0.3333333333333333 + B[2,4,9] * 0.16666666666666666) * B[2,1,3] + (B[2,1,9] * 0.16666666666666666 + B[2,4,9] * 0.3333333333333333) * B[2,4,3], (B[2,1,9] * 0.3333333333333333 + B[2,4,9] * 0.16666666666666666) * B[2,1,4] + (B[2,1,9] * 0.16666666666666666 + B[2,4,9] * 0.3333333333333333) * B[2,4,4], (B[2,1,9] * 0.3333333333333333 + B[2,4,9] * 0.16666666666666666) * B[2,1,5] + (B[2,1,9] * 0.16666666666666666 + B[2,4,9] * 0.3333333333333333) * B[2,4,5], (B[2,1,9] * 0.3333333333333333 + B[2,4,9] * 0.16666666666666666) * B[2,1,6] + (B[2,1,9] * 0.16666666666666666 + B[2,4,9] * 0.3333333333333333) * B[2,4,6], (B[2,1,9] * 0.3333333333333333 + B[2,4,9] * 0.16666666666666666) * B[2,1,7] + (B[2,1,9] * 0.16666666666666666 + B[2,4,9] * 0.3333333333333333) * B[2,4,7], (B[2,1,9] * 0.3333333333333333 + B[2,4,9] * 0.16666666666666666) * B[2,1,8] + (B[2,1,9] * 0.16666666666666666 + B[2,4,9] * 0.3333333333333333) * B[2,4,8], (B[2,1,9] * 0.3333333333333333 + B[2,4,9] * 0.16666666666666666) * B[2,1,9] + (B[2,1,9] * 0.16666666666666666 + B[2,4,9] * 0.3333333333333333) * B[2,4,9]}} + {{(B[3,1,1] * 0.3333333333333333 + B[3,4,1] * 0.16666666666666666) * B[3,1,1] + (B[3,1,1] * 0.16666666666666666 + B[3,4,1] * 0.3333333333333333) * B[3,4,1], (B[3,1,1] * 0.3333333333333333 + B[3,4,1] * 0.16666666666666666) * B[3,1,2] + (B[3,1,1] * 0.16666666666666666 + B[3,4,1] * 0.3333333333333333) * B[3,4,2], (B[3,1,1] * 0.3333333333333333 + B[3,4,1] * 0.16666666666666666) * B[3,1,3] + (B[3,1,1] * 0.16666666666666666 + B[3,4,1] * 0.3333333333333333) * B[3,4,3], (B[3,1,1] * 0.3333333333333333 + B[3,4,1] * 0.16666666666666666) * B[3,1,4] + (B[3,1,1] * 0.16666666666666666 + B[3,4,1] * 0.3333333333333333) * B[3,4,4], (B[3,1,1] * 0.3333333333333333 + B[3,4,1] * 0.16666666666666666) * B[3,1,5] + (B[3,1,1] * 0.16666666666666666 + B[3,4,1] * 0.3333333333333333) * B[3,4,5], (B[3,1,1] * 0.3333333333333333 + B[3,4,1] * 0.16666666666666666) * B[3,1,6] + (B[3,1,1] * 0.16666666666666666 + B[3,4,1] * 0.3333333333333333) * B[3,4,6], (B[3,1,1] * 0.3333333333333333 + B[3,4,1] * 0.16666666666666666) * B[3,1,7] + (B[3,1,1] * 0.16666666666666666 + B[3,4,1] * 0.3333333333333333) * B[3,4,7], (B[3,1,1] * 0.3333333333333333 + B[3,4,1] * 0.16666666666666666) * B[3,1,8] + (B[3,1,1] * 0.16666666666666666 + B[3,4,1] * 0.3333333333333333) * B[3,4,8], (B[3,1,1] * 0.3333333333333333 + B[3,4,1] * 0.16666666666666666) * B[3,1,9] + (B[3,1,1] * 0.16666666666666666 + B[3,4,1] * 0.3333333333333333) * B[3,4,9]}, {(B[3,1,2] * 0.3333333333333333 + B[3,4,2] * 0.16666666666666666) * B[3,1,1] + (B[3,1,2] * 0.16666666666666666 + B[3,4,2] * 0.3333333333333333) * B[3,4,1], (B[3,1,2] * 0.3333333333333333 + B[3,4,2] * 0.16666666666666666) * B[3,1,2] + (B[3,1,2] * 0.16666666666666666 + B[3,4,2] * 0.3333333333333333) * B[3,4,2], (B[3,1,2] * 0.3333333333333333 + B[3,4,2] * 0.16666666666666666) * B[3,1,3] + (B[3,1,2] * 0.16666666666666666 + B[3,4,2] * 0.3333333333333333) * B[3,4,3], (B[3,1,2] * 0.3333333333333333 + B[3,4,2] * 0.16666666666666666) * B[3,1,4] + (B[3,1,2] * 0.16666666666666666 + B[3,4,2] * 0.3333333333333333) * B[3,4,4], (B[3,1,2] * 0.3333333333333333 + B[3,4,2] * 0.16666666666666666) * B[3,1,5] + (B[3,1,2] * 0.16666666666666666 + B[3,4,2] * 0.3333333333333333) * B[3,4,5], (B[3,1,2] * 0.3333333333333333 + B[3,4,2] * 0.16666666666666666) * B[3,1,6] + (B[3,1,2] * 0.16666666666666666 + B[3,4,2] * 0.3333333333333333) * B[3,4,6], (B[3,1,2] * 0.3333333333333333 + B[3,4,2] * 0.16666666666666666) * B[3,1,7] + (B[3,1,2] * 0.16666666666666666 + B[3,4,2] * 0.3333333333333333) * B[3,4,7], (B[3,1,2] * 0.3333333333333333 + B[3,4,2] * 0.16666666666666666) * B[3,1,8] + (B[3,1,2] * 0.16666666666666666 + B[3,4,2] * 0.3333333333333333) * B[3,4,8], (B[3,1,2] * 0.3333333333333333 + B[3,4,2] * 0.16666666666666666) * B[3,1,9] + (B[3,1,2] * 0.16666666666666666 + B[3,4,2] * 0.3333333333333333) * B[3,4,9]}, {(B[3,1,3] * 0.3333333333333333 + B[3,4,3] * 0.16666666666666666) * B[3,1,1] + (B[3,1,3] * 0.16666666666666666 + B[3,4,3] * 0.3333333333333333) * B[3,4,1], (B[3,1,3] * 0.3333333333333333 + B[3,4,3] * 0.16666666666666666) * B[3,1,2] + (B[3,1,3] * 0.16666666666666666 + B[3,4,3] * 0.3333333333333333) * B[3,4,2], (B[3,1,3] * 0.3333333333333333 + B[3,4,3] * 0.16666666666666666) * B[3,1,3] + (B[3,1,3] * 0.16666666666666666 + B[3,4,3] * 0.3333333333333333) * B[3,4,3], (B[3,1,3] * 0.3333333333333333 + B[3,4,3] * 0.16666666666666666) * B[3,1,4] + (B[3,1,3] * 0.16666666666666666 + B[3,4,3] * 0.3333333333333333) * B[3,4,4], (B[3,1,3] * 0.3333333333333333 + B[3,4,3] * 0.16666666666666666) * B[3,1,5] + (B[3,1,3] * 0.16666666666666666 + B[3,4,3] * 0.3333333333333333) * B[3,4,5], (B[3,1,3] * 0.3333333333333333 + B[3,4,3] * 0.16666666666666666) * B[3,1,6] + (B[3,1,3] * 0.16666666666666666 + B[3,4,3] * 0.3333333333333333) * B[3,4,6], (B[3,1,3] * 0.3333333333333333 + B[3,4,3] * 0.16666666666666666) * B[3,1,7] + (B[3,1,3] * 0.16666666666666666 + B[3,4,3] * 0.3333333333333333) * B[3,4,7], (B[3,1,3] * 0.3333333333333333 + B[3,4,3] * 0.16666666666666666) * B[3,1,8] + (B[3,1,3] * 0.16666666666666666 + B[3,4,3] * 0.3333333333333333) * B[3,4,8], (B[3,1,3] * 0.3333333333333333 + B[3,4,3] * 0.16666666666666666) * B[3,1,9] + (B[3,1,3] * 0.16666666666666666 + B[3,4,3] * 0.3333333333333333) * B[3,4,9]}, {(B[3,1,4] * 0.3333333333333333 + B[3,4,4] * 0.16666666666666666) * B[3,1,1] + (B[3,1,4] * 0.16666666666666666 + B[3,4,4] * 0.3333333333333333) * B[3,4,1], (B[3,1,4] * 0.3333333333333333 + B[3,4,4] * 0.16666666666666666) * B[3,1,2] + (B[3,1,4] * 0.16666666666666666 + B[3,4,4] * 0.3333333333333333) * B[3,4,2], (B[3,1,4] * 0.3333333333333333 + B[3,4,4] * 0.16666666666666666) * B[3,1,3] + (B[3,1,4] * 0.16666666666666666 + B[3,4,4] * 0.3333333333333333) * B[3,4,3], (B[3,1,4] * 0.3333333333333333 + B[3,4,4] * 0.16666666666666666) * B[3,1,4] + (B[3,1,4] * 0.16666666666666666 + B[3,4,4] * 0.3333333333333333) * B[3,4,4], (B[3,1,4] * 0.3333333333333333 + B[3,4,4] * 0.16666666666666666) * B[3,1,5] + (B[3,1,4] * 0.16666666666666666 + B[3,4,4] * 0.3333333333333333) * B[3,4,5], (B[3,1,4] * 0.3333333333333333 + B[3,4,4] * 0.16666666666666666) * B[3,1,6] + (B[3,1,4] * 0.16666666666666666 + B[3,4,4] * 0.3333333333333333) * B[3,4,6], (B[3,1,4] * 0.3333333333333333 + B[3,4,4] * 0.16666666666666666) * B[3,1,7] + (B[3,1,4] * 0.16666666666666666 + B[3,4,4] * 0.3333333333333333) * B[3,4,7], (B[3,1,4] * 0.3333333333333333 + B[3,4,4] * 0.16666666666666666) * B[3,1,8] + (B[3,1,4] * 0.16666666666666666 + B[3,4,4] * 0.3333333333333333) * B[3,4,8], (B[3,1,4] * 0.3333333333333333 + B[3,4,4] * 0.16666666666666666) * B[3,1,9] + (B[3,1,4] * 0.16666666666666666 + B[3,4,4] * 0.3333333333333333) * B[3,4,9]}, {(B[3,1,5] * 0.3333333333333333 + B[3,4,5] * 0.16666666666666666) * B[3,1,1] + (B[3,1,5] * 0.16666666666666666 + B[3,4,5] * 0.3333333333333333) * B[3,4,1], (B[3,1,5] * 0.3333333333333333 + B[3,4,5] * 0.16666666666666666) * B[3,1,2] + (B[3,1,5] * 0.16666666666666666 + B[3,4,5] * 0.3333333333333333) * B[3,4,2], (B[3,1,5] * 0.3333333333333333 + B[3,4,5] * 0.16666666666666666) * B[3,1,3] + (B[3,1,5] * 0.16666666666666666 + B[3,4,5] * 0.3333333333333333) * B[3,4,3], (B[3,1,5] * 0.3333333333333333 + B[3,4,5] * 0.16666666666666666) * B[3,1,4] + (B[3,1,5] * 0.16666666666666666 + B[3,4,5] * 0.3333333333333333) * B[3,4,4], (B[3,1,5] * 0.3333333333333333 + B[3,4,5] * 0.16666666666666666) * B[3,1,5] + (B[3,1,5] * 0.16666666666666666 + B[3,4,5] * 0.3333333333333333) * B[3,4,5], (B[3,1,5] * 0.3333333333333333 + B[3,4,5] * 0.16666666666666666) * B[3,1,6] + (B[3,1,5] * 0.16666666666666666 + B[3,4,5] * 0.3333333333333333) * B[3,4,6], (B[3,1,5] * 0.3333333333333333 + B[3,4,5] * 0.16666666666666666) * B[3,1,7] + (B[3,1,5] * 0.16666666666666666 + B[3,4,5] * 0.3333333333333333) * B[3,4,7], (B[3,1,5] * 0.3333333333333333 + B[3,4,5] * 0.16666666666666666) * B[3,1,8] + (B[3,1,5] * 0.16666666666666666 + B[3,4,5] * 0.3333333333333333) * B[3,4,8], (B[3,1,5] * 0.3333333333333333 + B[3,4,5] * 0.16666666666666666) * B[3,1,9] + (B[3,1,5] * 0.16666666666666666 + B[3,4,5] * 0.3333333333333333) * B[3,4,9]}, {(B[3,1,6] * 0.3333333333333333 + B[3,4,6] * 0.16666666666666666) * B[3,1,1] + (B[3,1,6] * 0.16666666666666666 + B[3,4,6] * 0.3333333333333333) * B[3,4,1], (B[3,1,6] * 0.3333333333333333 + B[3,4,6] * 0.16666666666666666) * B[3,1,2] + (B[3,1,6] * 0.16666666666666666 + B[3,4,6] * 0.3333333333333333) * B[3,4,2], (B[3,1,6] * 0.3333333333333333 + B[3,4,6] * 0.16666666666666666) * B[3,1,3] + (B[3,1,6] * 0.16666666666666666 + B[3,4,6] * 0.3333333333333333) * B[3,4,3], (B[3,1,6] * 0.3333333333333333 + B[3,4,6] * 0.16666666666666666) * B[3,1,4] + (B[3,1,6] * 0.16666666666666666 + B[3,4,6] * 0.3333333333333333) * B[3,4,4], (B[3,1,6] * 0.3333333333333333 + B[3,4,6] * 0.16666666666666666) * B[3,1,5] + (B[3,1,6] * 0.16666666666666666 + B[3,4,6] * 0.3333333333333333) * B[3,4,5], (B[3,1,6] * 0.3333333333333333 + B[3,4,6] * 0.16666666666666666) * B[3,1,6] + (B[3,1,6] * 0.16666666666666666 + B[3,4,6] * 0.3333333333333333) * B[3,4,6], (B[3,1,6] * 0.3333333333333333 + B[3,4,6] * 0.16666666666666666) * B[3,1,7] + (B[3,1,6] * 0.16666666666666666 + B[3,4,6] * 0.3333333333333333) * B[3,4,7], (B[3,1,6] * 0.3333333333333333 + B[3,4,6] * 0.16666666666666666) * B[3,1,8] + (B[3,1,6] * 0.16666666666666666 + B[3,4,6] * 0.3333333333333333) * B[3,4,8], (B[3,1,6] * 0.3333333333333333 + B[3,4,6] * 0.16666666666666666) * B[3,1,9] + (B[3,1,6] * 0.16666666666666666 + B[3,4,6] * 0.3333333333333333) * B[3,4,9]}, {(B[3,1,7] * 0.3333333333333333 + B[3,4,7] * 0.16666666666666666) * B[3,1,1] + (B[3,1,7] * 0.16666666666666666 + B[3,4,7] * 0.3333333333333333) * B[3,4,1], (B[3,1,7] * 0.3333333333333333 + B[3,4,7] * 0.16666666666666666) * B[3,1,2] + (B[3,1,7] * 0.16666666666666666 + B[3,4,7] * 0.3333333333333333) * B[3,4,2], (B[3,1,7] * 0.3333333333333333 + B[3,4,7] * 0.16666666666666666) * B[3,1,3] + (B[3,1,7] * 0.16666666666666666 + B[3,4,7] * 0.3333333333333333) * B[3,4,3], (B[3,1,7] * 0.3333333333333333 + B[3,4,7] * 0.16666666666666666) * B[3,1,4] + (B[3,1,7] * 0.16666666666666666 + B[3,4,7] * 0.3333333333333333) * B[3,4,4], (B[3,1,7] * 0.3333333333333333 + B[3,4,7] * 0.16666666666666666) * B[3,1,5] + (B[3,1,7] * 0.16666666666666666 + B[3,4,7] * 0.3333333333333333) * B[3,4,5], (B[3,1,7] * 0.3333333333333333 + B[3,4,7] * 0.16666666666666666) * B[3,1,6] + (B[3,1,7] * 0.16666666666666666 + B[3,4,7] * 0.3333333333333333) * B[3,4,6], (B[3,1,7] * 0.3333333333333333 + B[3,4,7] * 0.16666666666666666) * B[3,1,7] + (B[3,1,7] * 0.16666666666666666 + B[3,4,7] * 0.3333333333333333) * B[3,4,7], (B[3,1,7] * 0.3333333333333333 + B[3,4,7] * 0.16666666666666666) * B[3,1,8] + (B[3,1,7] * 0.16666666666666666 + B[3,4,7] * 0.3333333333333333) * B[3,4,8], (B[3,1,7] * 0.3333333333333333 + B[3,4,7] * 0.16666666666666666) * B[3,1,9] + (B[3,1,7] * 0.16666666666666666 + B[3,4,7] * 0.3333333333333333) * B[3,4,9]}, {(B[3,1,8] * 0.3333333333333333 + B[3,4,8] * 0.16666666666666666) * B[3,1,1] + (B[3,1,8] * 0.16666666666666666 + B[3,4,8] * 0.3333333333333333) * B[3,4,1], (B[3,1,8] * 0.3333333333333333 + B[3,4,8] * 0.16666666666666666) * B[3,1,2] + (B[3,1,8] * 0.16666666666666666 + B[3,4,8] * 0.3333333333333333) * B[3,4,2], (B[3,1,8] * 0.3333333333333333 + B[3,4,8] * 0.16666666666666666) * B[3,1,3] + (B[3,1,8] * 0.16666666666666666 + B[3,4,8] * 0.3333333333333333) * B[3,4,3], (B[3,1,8] * 0.3333333333333333 + B[3,4,8] * 0.16666666666666666) * B[3,1,4] + (B[3,1,8] * 0.16666666666666666 + B[3,4,8] * 0.3333333333333333) * B[3,4,4], (B[3,1,8] * 0.3333333333333333 + B[3,4,8] * 0.16666666666666666) * B[3,1,5] + (B[3,1,8] * 0.16666666666666666 + B[3,4,8] * 0.3333333333333333) * B[3,4,5], (B[3,1,8] * 0.3333333333333333 + B[3,4,8] * 0.16666666666666666) * B[3,1,6] + (B[3,1,8] * 0.16666666666666666 + B[3,4,8] * 0.3333333333333333) * B[3,4,6], (B[3,1,8] * 0.3333333333333333 + B[3,4,8] * 0.16666666666666666) * B[3,1,7] + (B[3,1,8] * 0.16666666666666666 + B[3,4,8] * 0.3333333333333333) * B[3,4,7], (B[3,1,8] * 0.3333333333333333 + B[3,4,8] * 0.16666666666666666) * B[3,1,8] + (B[3,1,8] * 0.16666666666666666 + B[3,4,8] * 0.3333333333333333) * B[3,4,8], (B[3,1,8] * 0.3333333333333333 + B[3,4,8] * 0.16666666666666666) * B[3,1,9] + (B[3,1,8] * 0.16666666666666666 + B[3,4,8] * 0.3333333333333333) * B[3,4,9]}, {(B[3,1,9] * 0.3333333333333333 + B[3,4,9] * 0.16666666666666666) * B[3,1,1] + (B[3,1,9] * 0.16666666666666666 + B[3,4,9] * 0.3333333333333333) * B[3,4,1], (B[3,1,9] * 0.3333333333333333 + B[3,4,9] * 0.16666666666666666) * B[3,1,2] + (B[3,1,9] * 0.16666666666666666 + B[3,4,9] * 0.3333333333333333) * B[3,4,2], (B[3,1,9] * 0.3333333333333333 + B[3,4,9] * 0.16666666666666666) * B[3,1,3] + (B[3,1,9] * 0.16666666666666666 + B[3,4,9] * 0.3333333333333333) * B[3,4,3], (B[3,1,9] * 0.3333333333333333 + B[3,4,9] * 0.16666666666666666) * B[3,1,4] + (B[3,1,9] * 0.16666666666666666 + B[3,4,9] * 0.3333333333333333) * B[3,4,4], (B[3,1,9] * 0.3333333333333333 + B[3,4,9] * 0.16666666666666666) * B[3,1,5] + (B[3,1,9] * 0.16666666666666666 + B[3,4,9] * 0.3333333333333333) * B[3,4,5], (B[3,1,9] * 0.3333333333333333 + B[3,4,9] * 0.16666666666666666) * B[3,1,6] + (B[3,1,9] * 0.16666666666666666 + B[3,4,9] * 0.3333333333333333) * B[3,4,6], (B[3,1,9] * 0.3333333333333333 + B[3,4,9] * 0.16666666666666666) * B[3,1,7] + (B[3,1,9] * 0.16666666666666666 + B[3,4,9] * 0.3333333333333333) * B[3,4,7], (B[3,1,9] * 0.3333333333333333 + B[3,4,9] * 0.16666666666666666) * B[3,1,8] + (B[3,1,9] * 0.16666666666666666 + B[3,4,9] * 0.3333333333333333) * B[3,4,8], (B[3,1,9] * 0.3333333333333333 + B[3,4,9] * 0.16666666666666666) * B[3,1,9] + (B[3,1,9] * 0.16666666666666666 + B[3,4,9] * 0.3333333333333333) * B[3,4,9]}};
```
This expression could be reduced to something way shorter if the multiplications involving elements of B that are zero would be removed, and if the multiplication of elements of B times something would be replaced by that something.

This is of course done later by the backend, but I wonder if doing it in the frontend could spare us some time and memory in this case, where the large majority of those terms could be eliminated very early on.

**Comment by casella** [comment]
#11609 discusses a case where you still get expanded equations whose flat code is 14 million characters (!!). The problem is the same: reduction expressions applied to sparse matrix products, containing a lot of terms that are actually zero.

@perost please see #11609 for a proposed solution that would fix both tickets.
